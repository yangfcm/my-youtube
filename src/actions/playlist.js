import { defAxios as axios } from "../settings";
import { FETCH_PLAY_LIST, FETCH_PLAY_LIST_DETAIL, CATCH_ERROR } from "./types";

/** Fetch playlists under a particular channel or
 * the playlists under the channel of the authorized user
 */
export const fetchPlaylist = (pageToken, channelId, accessToken) => {
  return async (dispatch) => {
    try {
      let response;
      if (channelId) {
        response = await axios.get("/playlists", {
          params: {
            part: "snippet,contentDetails,status",
            maxResults,
            channelId,
            pageToken,
          },
        });
      }
      if (accessToken) {
        response = await axios.get("/playlists", {
          headers: {
            Authorization: accessToken,
          },
          params: {
            part: "snippet,contentDetails,status",
            maxResults,
            pageToken,
            mine: true,
          },
        });
      }
      dispatch({
        type: FETCH_PLAY_LIST,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: e.response.data,
      });
    }
  };
};

/** Fetch the videos in play list */
export const fetchPlaylistDetail = (playlistId, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/playlistItems", {
        params: {
          part: "snippet,contentDetails,status",
          maxResults: 8,
          playlistId,
          pageToken,
        },
      });
      dispatch({
        type: FETCH_PLAY_LIST_DETAIL,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: e.response.data,
      });
    }
  };
};
