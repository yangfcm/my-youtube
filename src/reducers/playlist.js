import { FETCH_PLAY_LIST, FETCH_PLAY_LIST_DETAIL } from "../actions/types";

export const playlistReducer = (state = {}, action) => {
  switch (action.type) {
    case FETCH_PLAY_LIST:
      return {
        ...state,
        playlist: action.payload,
      };
    case FETCH_PLAY_LIST_DETAIL:
      return {
        ...state,
        playlistDetail: action.payload,
      };
    default:
      return state;
  }
};
