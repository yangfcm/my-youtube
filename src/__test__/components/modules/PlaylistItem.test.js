import React from "react";
import { shallow } from "enzyme";
import { Link } from "react-router-dom";
import PlaylistItem from "components/modules/PlaylistItem";
import { playlistItemData } from "__test__/fixtures/playlist";

describe("Test PlaylistItem component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PlaylistItem playlist={playlistItemData} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct image", () => {
    expect(wrapper.find("img").prop("src")).toBe(
      playlistItemData.snippet.thumbnails.medium.url
    );
  });

  it("should render link and navigate to playlist detail page", () => {
    const links = wrapper.find(Link);
    expect(links.at(0).prop("to")).toBe(`/playlist/${playlistItemData.id}`);
    expect(links.at(1).prop("to")).toBe(`/playlist/${playlistItemData.id}`);
  });

  it("should render correct title", () => {
    expect(wrapper.find("div.card-body").text()).toContain(
      playlistItemData.snippet.title
    );
  });

  it("should display how many videos in playlist", () => {
    expect(wrapper.find("span.badge").text()).toContain(
      playlistItemData.contentDetails.itemCount
    );
  });
});
