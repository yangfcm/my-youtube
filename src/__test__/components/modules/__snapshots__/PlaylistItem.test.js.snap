// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PlaylistItem component should render component 1`] = `
<div
  className="card"
>
  <Link
    to="/playlist/PLOGi5-fAu8bH35xe7UG-nvk_Yf2Zp11Ds"
  >
    <img
      alt="TED Connects: Community and Hope"
      className="card-img-top"
      src="https://i.ytimg.com/vi/f0jETLKTW7w/mqdefault.jpg"
    />
  </Link>
  <div
    className="card-body"
  >
    <Link
      to="/playlist/PLOGi5-fAu8bH35xe7UG-nvk_Yf2Zp11Ds"
    >
      <p
        className="card-text"
      >
        TED Connects: Community and Hope
        Â 
        <br />
        <span
          className="badge badge-primary badge-pill"
        >
          15
           videos
        </span>
      </p>
    </Link>
  </div>
</div>
`;
